repub.go

```diff
@@ -167,23 +176,14 @@ func (rp *Republisher) Run(lastPublished cid.Cid) {
 		default:
 		}
 
-		// 2. If we have a value to publish, publish it now.
+		// 2. If there is a value to publish, publish it now.
 		if toPublish.Defined() {
-			for {
-				err := rp.pubfunc(rp.ctx, toPublish)
-				if err == nil {
-					break
-				}
-				// Keep retrying until we succeed or we abort.
-				// TODO(steb): We could try pulling new values
-				// off `update` but that's not critical (and
-				// complicates this code a bit). We'll pull off
-				// a new value on the next loop through.
-				select {
-				case <-time.After(rp.RetryTimeout):
-				case <-rp.ctx.Done():
-					return
-				}
+			err := rp.pubfunc(ctx, toPublish)
+			if err != nil {
+				// Keep retrying until publish succeeds, or run is stopped.
+				// Publish newer values if available.
+				longer.Reset(timeoutLong)
+				continue
```

Comment:
It seems we now don't "keep retrying" (we should reword that) and instead move on to the next loop (whether a new value to publish comes along superseding the failed one is not relevant). I would then suggest we refactor the next `WaitPub` (third) step into this one.

Before, the 1-2-3 succession implied all of them were always going to happen (at most step 3 would be delayed until the publish succeeded). Steps 2-3 should just be a single block now handling the fallout from the publish call: if successful, update `lastPublished` and trigger `waiter`, if not, do nothing and move on to the next loop (fine if that is not an actual big `if` and we keep the `continue` but it should more clearly reflect this).

Reply:
The logic does retry in that it resets the timer and continues the loop, retrying the pubfunc on the next timer expiration.  The previously failed value is retried unless a new value is provided.  Updated the comments to clarify.

Fixed the problem where this could lose a waiter if there were multiple calls to `WaitPublish()` during retry.  Added unit test for this.